src/commands.cc

```diff
@@ -342,6 +342,52 @@ const CommandDesc write_cmd = {
     write_buffer,
 };
 
+void force_write_buffer(const ParametersParser& parser, Context& context, const ShellContext&)
+{
+    Buffer& buffer = context.buffer();
+
+    if (parser.positional_count() == 0 and !(buffer.flags() & Buffer::Flags::File))
+        throw runtime_error("cannot write a non file buffer without a filename");
+
+    // if the buffer is in read-only mode and we try to save it directly
+    // or we try to write to it indirectly using e.g. a symlink, throw an error
+    if ((context.buffer().flags() & Buffer::Flags::ReadOnly)
+        and (parser.positional_count() == 0 or real_path(parser[0]) == buffer.name()))
+        throw runtime_error("cannot overwrite the buffer when in readonly mode");
+
+    auto filename = parser.positional_count() == 0 ?
+                        buffer.name() : parse_filename(parser[0]);
+
+    struct stat st;
+    if (::stat(filename.c_str(), &st) == 0)
+    {
+        const mode_t mode = st.st_mode | S_IWUSR;
+        if (::chmod(filename.c_str(), mode) < 0)
+            throw runtime_error("couldn't change file permissions");
+    }
+    else
+        st.st_mode = S_IRUSR | S_IWUSR | S_IRGRP | S_IROTH;
+
+    context.hooks().run_hook("BufWritePre", filename, context);
+    write_buffer_to_file(buffer, filename);
+    context.hooks().run_hook("BufWritePost", filename, context);
+
+    if (::chmod(filename.c_str(), st.st_mode) < 0)
```

Comment:
This should be done in an `on_scope_end`, as we want to restore the permissions even if something else fails.